kernel-rs/src/vm.rs

```diff
@@ -21,18 +21,454 @@ extern "C" {
+        if newsz < oldsz {
+            return Ok(oldsz);
+        }
+        oldsz = pgroundup(oldsz);
+        let mut a = oldsz;
+        while a < newsz {
+            let mem = kalloc() as *mut u8;
+            if mem.is_null() {
+                self.uvmdealloc(a, oldsz);
+                return Err(());
+            }
+            ptr::write_bytes(mem as *mut libc::CVoid, 0, PGSIZE);
+            if !self.mappages(a, PGSIZE, mem as usize, PTE_W | PTE_X | PTE_R | PTE_U) {
+                kfree(mem as *mut libc::CVoid);
+                self.uvmdealloc(a, oldsz);
+                return Err(());
+            }
+            a = a.wrapping_add(PGSIZE);
+        }
+        Ok(newsz)
+    }
+
+    /// Deallocate user pages to bring the process size from oldsz to
+    /// newsz.  oldsz and newsz need not be page-aligned, nor does newsz
+    /// need to be less than oldsz.  oldsz can be larger than the actual
+    /// process size.  Returns the new process size.
+    pub unsafe fn uvmdealloc(&mut self, oldsz: usize, newsz: usize) -> usize {
+        if newsz >= oldsz {
+            return oldsz;
+        }
+        let newup: usize = pgroundup(newsz);
+        if newup < pgroundup(oldsz) {
+            self.uvmunmap(newup, oldsz.wrapping_sub(newup), 1);
+        }
+        newsz
+    }
+
+    /// Free user memory pages,
+    /// then free page-table pages.
+    pub unsafe fn uvmfree(&mut self, sz: usize) {
+        self.uvmunmap(0, sz, 1);
+        self.freewalk();
+    }
+
+    /// Given a parent process's page table, copy
+    /// its memory into a child's page table.
+    /// Copies both the page table and the
+    /// physical memory.
+    /// Returns Ok(()) on success, Err(()) on failure.
+    /// Frees any allocated pages on failure.
+    pub unsafe fn uvmcopy(&mut self, mut new: &mut RawPageTable, sz: usize) -> Result<(), ()> {
+        for i in num_iter::range_step(0, sz, PGSIZE) {
+            let pte_op = walk(self, i, 0);
+            if pte_op.is_none() {
+                panic!("uvmcopy: pte should exist");
+            }
+            let pte = pte_op.unwrap();
+            if !pte.check_flag(PTE_V) {
+                panic!("uvmcopy: page not present");
+            }
+            let mut new_ptable = scopeguard::guard(new, |ptable| {
+                ptable.uvmunmap(0, i, 1);
+            });
+            let pa = pte.get_pa();
+            let flags = pte.get_flags() as u32;
+            let mem = kalloc() as *mut u8;
+            if mem.is_null() {
+                return Err(());
+            }
+            ptr::copy(
+                pa as *mut u8 as *const libc::CVoid,
+                mem as *mut libc::CVoid,
+                PGSIZE,
+            );
+            if !(*new_ptable).mappages(i, PGSIZE, mem as usize, flags as i32) {
+                kfree(mem as *mut libc::CVoid);
+                return Err(());
+            }
+            new = scopeguard::ScopeGuard::into_inner(new_ptable);
```

Comment:
`use scopeguard::ScopeGuard;` 추가해 주세요..
```suggestion
            new = ScopeGuard::into_inner(new_ptable);
```